src/main/scala/com/bazaarvoice/sswf/model/StepsHistory.scala

```diff
@@ -0,0 +1,171 @@
+package com.bazaarvoice.sswf.model
+
+import java.{util, lang}
+import java.util.Collections
+
+import com.amazonaws.services.simpleworkflow.model.EventType._
+import com.amazonaws.services.simpleworkflow.model._
+import com.bazaarvoice.sswf.StepEventState._
+import com.bazaarvoice.sswf.{InputParser, SSWFStep, StepEventState}
+import org.joda.time.{DateTime, Duration}
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+import scala.reflect.{ClassTag, classTag}
+
+object WorkflowEventToken {
+  override def toString = "WORKFLOW"
+}
+
+case class StepEvent[StepEnum <: (Enum[StepEnum] with SSWFStep) : ClassTag](id: Long,
+                                                                            event: Either[StepEnum, WorkflowEventToken.type],
+                                                                            result: String,
+                                                                            state: StepEventState,
+                                                                            start: DateTime,
+                                                                            end: Option[DateTime],
+                                                                            cumulativeActivityTime: Duration)
+
+case class StepsHistory[SSWFInput, StepEnum <: (Enum[StepEnum] with SSWFStep) : ClassTag](input: SSWFInput, events: java.util.List[StepEvent[StepEnum]], firedTimers: java.util.Set[StepEnum])
+
+object HistoryFactory {
+  def from[SSWFInput, StepEnum <: (Enum[StepEnum] with SSWFStep) : ClassTag](swfHistory: List[HistoryEvent],
+                                                                             inputParser: InputParser[SSWFInput]): StepsHistory[SSWFInput, StepEnum] = {
+
+    def enumFromName(name: String): StepEnum = Enum.valueOf(classTag[StepEnum].runtimeClass.asInstanceOf[Class[StepEnum]], name)
+
+    def filterStarts(s: List[StepEvent[StepEnum]]): List[StepEvent[StepEnum]] = s match {
```

Comment:
Just kidding. Although there is a bug here. Good call. I wanted to keep only the last event for a given activityScheduledId. After the activity is scheduled, it goes through shedule-to-start-timeout or start, and then start-to-finish-timeout, complete, or failed. I'm basically compressing these so you only see the final event in the sequence.

If the event hasn't been started, it will show as scheduled. If it has, it will show as started _only_. If it's done, it will show as completed _only_.

Like I said, there is a bug, which may have contributed to the confusion. I'll fix it.